### DISCUSSION TITLE
Advanced inference engine features

### DISCUSSION THREAD

**Discussion by kir-gadjello** [EricLBuehler/mistral.rs#245] [Ideas]
Advanced inference engine features I think many will find desirable:

1. Speculative decoding with fast ngram-like or small transformer-like models (without requirement to have exact same tokenizers): https://github.com/vllm-project/vllm/pull/2188
2. Content-aware grammar decoding: when inside a markdown code block use the corresponding programming language grammar. I think all major LLM providers might use variants of this. Also have json mode (+ json schema mode like in tabbyapi and llama.cpp) if you don't have any yet, these are useful.
3. KV-cache compression with Hadamard transform as in latest exllama v2: https://github.com/turboderp/exllamav2/commit/324404ebe4e3c4dd0447ffc1290c312de1df02be
4. Smart KV-cache offload into RAM (could be probably combined with https://www.semanticscholar.org/paper/Scissorhands%3A-Exploiting-the-Persistence-of-for-LLM-Liu-Desai/d6eeb2898bd9bd34744194ef543062dda6c4531a but this is research territory).
4. Support for at least one strong vision-language model: https://huggingface.co/openbmb/MiniCPM-V-2 https://huggingface.co/OpenGVLab/InternVL-Chat-V1-5 with an option to compute visual frontend model on CPU. You might find it easier to ship visual transformer part via onnx.
5. Dynamic per-request yarn rope-scaling, if it is possible.
6. Advanced multi-slot inference (ability to extract batch parallelism from a few concurrent completion requests, I know it might be hard but would be also interesting if there were RadixAttention impl like in https://github.com/sgl-project/sglang - this all is useful when you have an agent system that produces many concurrent api calls with a shared prefix/kvcache)
7. Multi-model inference with minimal VRAM<->RAM swapping: there is a usecase where a person needs to run a strong instruction-tuned model (takes almost all VRAM, used intermittently) concurrently with a smaller but sill considerable code completion model. It would be very convenient if the engine were able to run both at the same time assuming bursty exclusive usage pattern and minimizing latency with preference given to one or the other model.
8. Support for Medusa or EAGLE with heads like these: https://huggingface.co/text-generation-inference
9. High-quality calibrated quantization of long-context models (32, 64, 128k).
10. Per-token embedding output (choose one of the last layers) for classifiers.

Research Territory:

1. Smart KV-cache compression (leave salient tokens alone)
2. Support for the "best" current quantization format for running 70B-tier models on 24GB GPUs - might be AQLM, iQuant or another scheme.

Thanks for hearing me out!

**Comment by lucasavila00** [comment]
@kir-gadjello thanks for this. I really appreciate it.

> Content-aware grammar decoding: when inside a markdown code block use the corresponding programming language grammar. I think all major LLM providers might use variants of this. Also have json mode (+ json schema mode like in tabbyapi and llama.cpp) if you don't have any yet, these are useful.

mistral.rs supports regex and yacc based grammars so a big part of it can be implemented "in userland". I was hoping to build something that translates json schema to a yacc grammar automatically. It's not too different from how outlines do it.

---

You mention both different quants and `RadixAttention`. These are usually hard to combine efficiently. Mistral.rs currently has no paging for KV cache, meaning we can use quant kernels without a lot of changes.

Should mistral.rs add paging, we'd need to change all attention kernels to be aware of the memory paging.

An option would be to copy the paging implementation from sglang or vllm. But then we would be limited to using kernels from either project, because the kernels must be aware of the KV paging

**Comment by EricLBuehler** [comment]
@kir-gadjello, thank you for listing these out! I have gone through each point and responded to each.

> You mention both different quants and RadixAttention. These are usually hard to combine efficiently. Mistral.rs currently has no paging for KV cache, meaning we can use quant kernels without a lot of changes.
> 
> Should mistral.rs add paging, we'd need to change all attention kernels to be aware of the memory paging.
> 
> An option would be to copy the paging implementation from sglang or vllm. But then we would be limited to using kernels from either project, because the kernels must be aware of the KV paging

@lucasavila00 can you please elaborate on how different quants and `RadixAttention` are hard to combine?

- **Speculative decoding with fast ngram-like or small transformer-like models (without requirement to have exact same tokenizers): https://github.com/vllm-project/vllm/pull/2188**
This is being implemented in #242.
- **Content-aware grammar decoding: when inside a markdown code block use the corresponding programming language grammar. I think all major LLM providers might use variants of this. Also have json mode (+ json schema mode like in tabbyapi and llama.cpp) if you don't have any yet, these are useful.**
I think we can provide specialized grammars to implement this: where it only activates if in a code block perhaps.
- [ ] Specialized grammars for context aware decoding
- [ ] Default JSON schema
- **KV-cache compression with Hadamard transform as in latest exllama v2: https://github.com/turboderp/exllamav2/commit/324404ebe4e3c4dd0447ffc1290c312de1df02be**
Can you please elaborate what this does?
- [ ] Implement KV cache quantization or compression
- **Smart KV-cache offload into RAM (could be probably combined with https://www.semanticscholar.org/paper/Scissorhands%3A-Exploiting-the-Persistence-of-for-LLM-Liu-Desai/d6eeb2898bd9bd34744194ef543062dda6c4531a but this is research territory).**
- [ ] Implement KV cache offloading
- **Support for at least one strong vision-language model: https://huggingface.co/openbmb/MiniCPM-V-2 https://huggingface.co/OpenGVLab/InternVL-Chat-V1-5 with an option to compute visual frontend model on CPU. You might find it easier to ship visual transformer part via onnx.**
We plan to implement vision models this week. 
- **Dynamic per-request yarn rope-scaling, if it is possible.**
How would this work? We support the Yarn rope scaling method already:
> https://github.com/EricLBuehler/mistral.rs/blob/d442e0202bd19682e3e8efeca151362046b9fc33/mistralrs-core/src/layers.rs#L70-L73
- **Advanced multi-slot inference (ability to extract batch parallelism from a few concurrent completion requests, I know it might be hard but would be also interesting if there were RadixAttention impl like in https://github.com/sgl-project/sglang - this all is useful when you have an agent system that produces many concurrent api calls with a shared prefix/kvcache)**
We currently implement prefix caching. Do you mean cross-gpu inference support by multi-slot inference? Please let me know if I am misunderstanding.
- **Multi-model inference with minimal VRAM<->RAM swapping: there is a usecase where a person needs to run a strong instruction-tuned model (takes almost all VRAM, used intermittently) concurrently with a smaller but sill considerable code completion model. It would be very convenient if the engine were able to run both at the same time assuming bursty exclusive usage pattern and minimizing latency with preference given to one or the other model.**
I'm designing the speculative sampling to support dual-pipelines generally, so this can be implemented with that. 
- **Support for Medusa or EAGLE with heads like these: https://huggingface.co/text-generation-inference**
I think this is a very interesting topic and again can be implemented from something like the speculative decoding multiple-pipeline architecture.
- **High-quality calibrated quantization of long-context models (32, 64, 128k).**
Can you please link a paper or an issue for this? I think this sounds good, though.
- **Per-token embedding output (choose one of the last layers) for classifiers.**
Can you please link a paper or an issue for this? Which classifier would benefit from this?

**Reply by oldgithubman** [reply]
> * **KV-cache compression with Hadamard transform as in latest exllama v2: [turboderp/exllamav2@324404e](https://github.com/turboderp/exllamav2/commit/324404ebe4e3c4dd0447ffc1290c312de1df02be)**
>   Can you please elaborate what this does?

Quantizes the KV cache with the goal of using less space. llama.cpp allows you to specify quantization types for K and V individually. Ideally, we'd be able to use all the same quantization types as gguf. If you're asking about "Hadamard transform," I don't know either.

The ability to have multiple models loaded simultaneously would be useful (for example, for continue.dev, a coding model and an embedding model).

> High-quality calibrated quantization of long-context models (32, 64, 128k).
Can you please link a paper or an issue for this? I think this sounds good, though.

I'm not sure what he meant by this either, but I too like the sound of it

**Reply by EricLBuehler** [reply]
Yeah, these are all super interesting. I think KV cache compression is the most realistic to implement right now, I'm not really sure what the calibrated quantization would mean - perhaps something like the `IQ...` quants?

**Reply by oldgithubman** [reply]
> Yeah, these are all super interesting. I think KV cache compression is the most realistic to implement right now, I'm not really sure what the calibrated quantization would mean - perhaps something like the `IQ...` quants?

Maybe. IQ quants are nice. I almost always use them

**Reply by EricLBuehler** [reply]
Sounds good. I will add those soon!

**Comment by lucasavila00** [comment]
> @lucasavila00 can you please elaborate on how different quants and RadixAttention are hard to combine?

Paged attention or radix attention work by storing a bunch of slots for the KV cache of each token. Let's call it the KV array.

At inference time, the model receives a list of indexes as the KV cache, not tensors with the data itself.

Then, at inference time, the attention kernels dynamically use the index to access the item of the KV array and do the attention calculation. It never creates a contiguous tensor with the KV cache.

So if we get the AWQ kernels from https://github.com/casper-hansen/AutoAWQ_kernels which work on contiguous tensors, we can't use them with paged attention without rewriting the attention part to work with the index indirection.

**Reply by lucasavila00** [reply]
We can create the contigous tensor on demand. But that uses twice as much RAM, defeating the purpose of PA. And also it's slow, defeating the purpose of quantization. So it's hard to support both at once efficiently

**Reply by EricLBuehler** [reply]
Thanks for explaining! Do you see this sort of paging as a potential performance gain for prefix caching? Currently, we need to clone and cat in/out the entire KV cache of the model per completion and prompt step. I see that as a weakness, but I'm not sure if it is significant enough to warrant changing our attention implementation.

**Reply by lucasavila00** [reply]
That's a hard question.

It depends on what we want to use mistral.rs as.

If we want it to fill in the shoes of vLLM as a production server thing, then PA is likely required, as it reduces RAM usage and provides better almost-free caching and prefix-caching (sglang's implementation that vLLM is copying)

But if we want it to fill in the shoes of llama.cpp as a way of running local models than the current approach is better. We can re-use code from other projects easier, likely faster due to no indirection and so on. (Although llama.cpp is also adding paging, but AFAIK not ready)

>  Currently, we need to clone and cat in/out the entire KV cache of the model per completion and prompt step.

A "smarter scheduler" or engine loop can likely avoid these copies if the next step does not require it. Kind of similar to that completion skip of https://github.com/EricLBuehler/mistral.rs/pull/209

**Reply by lucasavila00** [reply]
My current thoughts are that ideally mistral.rs should abstract more of the core loop such that it is possible to have PA, but PA doesn't need to work with all other features combined.

So let's say I want to create just AWQ quantization with PA for the llama architecture. Currently that is not possible because the engine deals with KV cache, I cannot just implement the Pipeline trait.

**Reply by lucasavila00** [reply]
Perhaps the cache should be handled by another KvManager trait. And then it's left to the user to know which KvManagers can be used with each Pipelines, or something like that.

**Reply by lucasavila00** [reply]
:thinking: 

The scheduler also needs to be cache aware, so I guess it should be a KVManagerScheduler trait?

**Reply by EricLBuehler** [reply]
Hmm, that could work. #242 is going to bring some major changes internally to abstract the sampling and generation process overall. Perhaps we could include these traits in a PR later.

**Reply by EricLBuehler** [reply]
> The scheduler also needs to be cache aware, so I guess it should be a KVManagerScheduler trait?

Does it need to be cache aware? That would be an optimization, to reduce the clone+cat steps which we could implement right now.

**Reply by lucasavila00** [reply]
It does IMO. Let's say we got 2 requests, but can only execute 1 at a time because of configurations.

One request is 90% prefix-cached, the other 0% prefix-cached.

If we run the 0% prefix-cached we'll delete the cache and then we'll need to start the 90% req from scratch.

If we run the 90% prefix-cached first we can increase cache re-use. Sglang does this, vLLM does not. vLLm is adding this ability though.

**Reply by EricLBuehler** [reply]
Yes, we should prioritize requests that way.

**Comment by EricLBuehler** [comment]
@kir-gadjello @lucasavila00 we now have PagedAttention!
